Trying to debug a failure in the backup/schemachange tests. I'm leaving notes to return to later.

It's hitting this error:
```
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_add_check_udf/post_commit_stage_1_of_3/restore_all_tables_in_database
    backup.go:575: error executing query="DROP DATABASE IF EXISTS \"defaultdb\" CASCADE" args=[]: pq: referenced descriptor ID 1: looking up ID 1: descriptor not found
```

It's linked to the SHOW BACKUP FROM LATEST query that's made by the test here: https://github.com/cockroachdb/cockroach/blob/0580abda806e0b196025e18047bfd3202862abdd/pkg/sql/schemachanger/sctest/backup.go#L341-L345

I added additional logging to show the stack when the error happens:
```
not found: goroutine 8923 [running]:
runtime/debug.Stack()
	GOROOT/src/runtime/debug/stack.go:26 +0x64
github.com/cockroachdb/cockroach/pkg/util/debugutil.Stack(...)
	pkg/util/debugutil/debugutil.go:63
github.com/cockroachdb/cockroach/pkg/sql/catalog.NewDescriptorNotFoundError(0x1)
	pkg/sql/catalog/errors.go:92 +0x20
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.requiredError({0x1082a2007?, 0x0?}, 0x1)
	pkg/sql/catalog/internal/catkv/catalog_query.go:219 +0x2e8
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.build({0x1082a2007, 0x3}, 0x1, 0x0, 0x1)
	pkg/sql/catalog/internal/catkv/catalog_query.go:189 +0x2b8
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.catalogQuery.processDescriptorResultRow({{{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, 0x1, {0x1082a2007, 0x3}}, {{0x14005ed9c40?, 0x10ea99b40?, ...}, ...}, ...)
	pkg/sql/catalog/internal/catkv/catalog_query.go:116 +0x90
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.catalogQuery.query({{{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, 0x1, {0x1082a2007, 0x3}}, {0x10a693da0, 0x14006dd9ae0}, ...)
	pkg/sql/catalog/internal/catkv/catalog_query.go:71 +0x264
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.catalogReader.GetByIDs({{{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}}, {0x10a693da0, 0x14006dd9ae0}, 0x14003167810, {0x14005ed9c04, 0x1, ...}, ...)
	pkg/sql/catalog/internal/catkv/catalog_reader.go:419 +0x194
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.(*cachedCatalogReader).GetByIDs(0x1400594be60, {0x10a693da0, 0x14006dd9ae0}, 0x14003167810, {0x14005ed9c04, 0x1, 0x1}, 0x1, {0x1082a2007, 0x3}, ...)
	pkg/sql/catalog/internal/catkv/catalog_reader_cached.go:387 +0x250
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.mustGetDescriptorByIDs({0x14001ef08c0, {0x10a693ef0, 0x14003fa1200}, 0x14003a7c8c0, 0x1400226a000, {{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, ...}, ...)
	pkg/sql/catalog/lease/storage.go:453 +0xf4
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.mustGetDescriptorByID({0x14001ef08c0, {0x10a693ef0, 0x14003fa1200}, 0x14003a7c8c0, 0x1400226a000, {{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, ...}, ...)
	pkg/sql/catalog/lease/storage.go:439 +0x88
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.getForExpiration.func1({0x10a693da0, 0x14006dd9ae0}, 0x14003167810)
	pkg/sql/catalog/lease/storage.go:411 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0x14003167810, {0x10a693da0, 0x14006dd9ae0}, 0x14004cd3c78)
	pkg/kv/txn.go:1139 +0x74
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x10a693da0, 0x14006dd9ae0}, 0x14003167810, 0x1?)
	pkg/kv/db.go:1114 +0x3c
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x8?, {0x10a693da0, 0x14006dd9ae0}, 0x5?, 0x0?, 0x0, 0x14004cd3c78)
	pkg/kv/db.go:1077 +0xa8
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(...)
	pkg/kv/db.go:1052
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.storage.getForExpiration({0x14001ef08c0, {0x10a693ef0, 0x14003fa1200}, 0x14003a7c8c0, 0x1400226a000, {{0x10f07e100, 0x10f07e120, {0x1}}, {0x10f07e100}, 0x0}, ...}, ...)
	pkg/sql/catalog/lease/storage.go:405 +0x94
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).readOlderVersionForTimestamp(0x14001f33088, {0x10a693da0, 0x14006dd9ae0}, 0x1, {0x14006170600?, 0x1e03cde0?})
	pkg/sql/catalog/lease/lease.go:970 +0x19c
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).Acquire(0x14001f33088, {0x10a693da0, 0x14006dd9ae0}, {0x10a66b460, 0x14006170600}, 0x1)
	pkg/sql/catalog/lease/lease.go:2201 +0x118
github.com/cockroachdb/cockroach/pkg/sql/regions.NewCachedDatabaseRegionsAt({0x10a693da0, 0x14006dd9ae0}, 0x1400442a600?, 0x14001f33088, {0x14006dd9ae0?, 0x3a34840?})
	pkg/sql/regions/cached_db_regions.go:35 +0xc8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).checkDescriptorTwoVersionInvariant(0x14004ee4f08, {0x10a693da0, 0x14006dd9ae0})
	pkg/sql/conn_executor_exec.go:2275 +0x84
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).commitSQLTransactionInternal(0x14004ee4f08, {0x10a693da0?, 0x14006dd9ae0?})
	pkg/sql/conn_executor_exec.go:2457 +0x3a0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).commitSQLTransaction(0x14004ee4f08, {0x10a693da0, 0x14006dd9ae0}, {0x10a6c0940, 0x14007bca3a0}, 0x14004cd44e8)
	pkg/sql/conn_executor_exec.go:2339 +0x19c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).handleAutoCommit(0x14004ee4f08, {0x10a693da0, 0x14006dd9ae0}, {0x10a6c0940, 0x14007bca3a0})
	pkg/sql/conn_executor_exec.go:4129 +0xec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenStateWithPausablePortal.func12({0x10a693da0?, 0x14006dd9ae0?})
	pkg/sql/conn_executor_exec.go:1618 +0xd4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenStateWithPausablePortal(0x14004ee4f08, {0x10a693da0, 0x14006dd9a90}, {{0x10a6c0940, 0x14007bca3a0}, {0x0, 0x0, 0x0}, {0x14002f16f2c, 0x2b}, ...}, ...)
	pkg/sql/conn_executor_exec.go:2149 +0x3818
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x10a693da0?, 0x14006dd9a90?})
	pkg/sql/conn_executor_exec.go:171 +0x6c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x14004ee4f08, {0x10a693da0, 0x14006dd9a90}, {0x10a6c0940, 0x14007bca3a0}, 0x0, 0x14004cd5670)
	pkg/sql/conn_executor_exec.go:4470 +0x310
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x14004ee4f08, {0x10a693da0, 0x14006dd9a90}, {{0x10a6c0940, 0x14007bca3a0}, {0x0, 0x0, 0x0}, {0x14002f16f2c, 0x2b}, ...}, ...)
	pkg/sql/conn_executor_exec.go:170 +0x508
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(...)
	pkg/sql/conn_executor.go:2426
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x14004ee4f08)
	pkg/sql/conn_executor.go:2431 +0x1948
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x14004ee4f08, {0x10a6947d0, 0x14006798f00}, 0x30000007d?, 0x140078ffc78?, 0x14001fa5c00?)
	pkg/sql/conn_executor.go:2339 +0x158
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0x14004d403c0?, {0x10a6947d0?, 0x14006798f00?}, {0x14001330a20?}, 0x2?, 0x0?)
	pkg/sql/conn_executor.go:1118 +0xa0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0x1400434e008, {0x10a6947d0, 0x14006798f00}, {0x0, 0x4, {0x1, {0x1082a1f92, 0x3}, {0x1400739d960, 0xf}, ...}, ...}, ...)
	pkg/sql/pgwire/conn.go:253 +0x304
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4()
	pkg/sql/pgwire/server.go:1260 +0xa4
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl in goroutine 8921
	pkg/sql/pgwire/server.go:1257 +0x424
```

I'm not sure if the `readOlderVersionForTimestamp` is expected above. This issue goes away when I remove the two new calls I added to lookup the system tables (`txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleOptionsTableID)` and `txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.DatabaseRoleSettingsTableID)`)

--- a/pkg/sql/information_schema.go
+++ b/pkg/sql/information_schema.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/schemadesc"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr"
 	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/rolemembershipcache"
 	"github.com/cockroachdb/cockroach/pkg/sql/roleoption"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/builtins"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtinsregistry"
@@ -2995,6 +2996,8 @@ GROUP BY
 			}
 			return nil
 		},
+		rolemembershipcache.CheckRoleOptionsVersion(),
+		rolemembershipcache.CheckDatabaseRoleSettingsVersion(),
 	); err != nil {
 		return err
 	}

--- a/pkg/sql/logictest/testdata/logic_test/pg_catalog
+++ b/pkg/sql/logictest/testdata/logic_test/pg_catalog
@@ -4954,7 +4954,7 @@ WHERE usename IN ('testuser1', 'testuser2')
 ORDER BY usename
 ----
 usename    usecreatedb  useconfig
-testuser1  false        NULL
+testuser1  true         {timezone=America/Los_Angeles,application_name=a}
 testuser2  false        NULL
 
 query TBTBBT colnames
@@ -4963,10 +4963,10 @@ FROM pg_roles
 WHERE rolname IN ('testuser1', 'testuser2', 'testrole1')
 ORDER BY rolname
 ----
-rolname    rolcreatedb  rolconfig  rolinherit  rolcanlogin  rolvaliduntil
-testrole1  false        NULL       true        false        NULL
-testuser1  true         NULL       true        true         NULL
-testuser2  false        NULL       true        true         NULL
+rolname    rolcreatedb  rolconfig                                          rolinherit  rolcanlogin  rolvaliduntil
+testrole1  false        NULL                                               true        false        NULL
+testuser1  true         {timezone=America/Los_Angeles,application_name=a}  true        true         NULL
+testuser2  false        NULL                                               true        true         3022-01-01 00:00:00 +0000 UTC
 
 query TBBBBT colnames
 SELECT rolname, rolcreatedb, rolcreaterole, rolinherit, rolcanlogin, rolvaliduntil
@@ -4978,7 +4978,7 @@ rolname    rolcreatedb  rolcreaterole  rolinherit  rolcanlogin  rolvaliduntil
 root       true         true           true        true         NULL
 testrole1  false        false          true        false        NULL
 testuser1  true         false          true        true         NULL
-testuser2  false        true           true        true         NULL
+testuser2  false        true           true        true         3022-01-01 00:00:00 +0000 UTC
 
 # Testing users that have admin role
 

--- a/pkg/sql/pg_catalog.go
+++ b/pkg/sql/pg_catalog.go
@@ -47,6 +47,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/volatility"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
+	"github.com/cockroachdb/cockroach/pkg/sql/syntheticprivilege"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/sql/vtable"
 	"github.com/cockroachdb/cockroach/pkg/util/collatedstring"
@@ -607,24 +608,58 @@ func userIsSuper(
 	return tree.DBool(isSuper), err
 }
 
+// userHasReplicationPrivilegeOrRoleOption checks if a user has replication
+// capability either from their role options or from global synthetic
+// privileges. The role option check is done first as it's a cheap in-memory
+// check, and the global privilege check is only done if needed.
 func userHasReplicationPrivilegeOrRoleOption(
-	ctx context.Context, p *planner, userName username.SQLUsername,
-) (tree.DBool, error) {
-	replication, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.REPLICATION, userName)
+	ctx context.Context, p *planner, userName username.SQLUsername, options roleOptions,
+) (bool, error) {
+	ok, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.REPLICATION)
 	if err != nil {
-		return *tree.DBoolFalse, err
+		return false, err
+	} else if ok {
+		return true, nil
 	}
+	ok, err = userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.REPLICATIONDEST)
+	if err != nil {
+		return false, err
+	} else if ok {
+		return true, nil
+	}
+	ok, err = userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.REPLICATIONSOURCE)
+	if err != nil {
+		return false, err
+	} else if ok {
+		return true, nil
+	}
+	return false, nil
+}
 
-	replicationDest, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.REPLICATIONDEST, userName)
+// userHasPrivilegeOrRoleOption checks if a user has a privilege either from
+// their role options (from the roleOptionsJSON) or from a global synthetic
+// privilege. The roleOption check is done first as it's a cheap in-memory
+// check, and the global privilege check is only done if needed.
+func userHasPrivilegeOrRoleOption(
+	ctx context.Context,
+	p *planner,
+	userName username.SQLUsername,
+	options roleOptions,
+	priv privilege.Kind,
+) (bool, error) {
+	roleOptionKey := string(priv.InternalKey())
+	hasRoleOption, err := options.Exists(roleOptionKey)
 	if err != nil {
-		return *tree.DBoolFalse, err
+		return false, err
 	}
-
-	replicationSrc, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.REPLICATIONSOURCE, userName)
+	if hasRoleOption {
+		return true, nil
+	}
+	hasPriv, err := p.HasPrivilege(ctx, syntheticprivilege.GlobalPrivilegeObject, priv, userName)
 	if err != nil {
-		return *tree.DBoolFalse, err
+		return false, err
 	}
-	return tree.DBool(replication || replicationDest || replicationSrc), nil
+	return hasPriv, nil
 }
 
 var pgCatalogAuthIDTable = virtualSchemaTable{
@@ -649,17 +684,17 @@ https://www.postgresql.org/docs/9.5/catalog-pg-authid.html`,
 				return err
 			}
 
-			bypassRLS, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.BYPASSRLS, userName)
+			bypassRLS, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.BYPASSRLS)
 			if err != nil {
 				return err
 			}
 
-			createRole, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.CREATEROLE, userName)
+			createRole, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.CREATEROLE)
 			if err != nil {
 				return err
 			}
 
-			createDB, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.CREATEDB, userName)
+			createDB, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.CREATEDB)
 			if err != nil {
 				return err
 			}
@@ -669,7 +704,7 @@ https://www.postgresql.org/docs/9.5/catalog-pg-authid.html`,
 				return err
 			}
 
-			replication, err := userHasReplicationPrivilegeOrRoleOption(ctx, p, userName)
+			replication, err := userHasReplicationPrivilegeOrRoleOption(ctx, p, userName, options)
 			if err != nil {
 				return err
 			}
@@ -682,7 +717,7 @@ https://www.postgresql.org/docs/9.5/catalog-pg-authid.html`,
 				tree.MakeDBool(isRoot || tree.DBool(createRole)), // rolcreaterole
 				tree.MakeDBool(isRoot || tree.DBool(createDB)),   // rolcreatedb
 				tree.MakeDBool(roleCanLogin),                     // rolcanlogin.
-				tree.MakeDBool(replication),                      // rolreplication
+				tree.MakeDBool(tree.DBool(replication)),          // rolreplication
 				tree.MakeDBool(tree.DBool(bypassRLS)),            // rolbypassrls
 				negOneVal,                                        // rolconnlimit
 				passwdStarString,                                 // rolpassword
@@ -3028,17 +3063,17 @@ https://www.postgresql.org/docs/9.5/view-pg-roles.html`,
 					return err
 				}
 
-				bypassRLS, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.BYPASSRLS, userName)
+				bypassRLS, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.BYPASSRLS)
 				if err != nil {
 					return err
 				}
 
-				createRole, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.CREATEROLE, userName)
+				createRole, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.CREATEROLE)
 				if err != nil {
 					return err
 				}
 
-				createDB, err := p.UserHasGlobalPrivilegeOrRoleOption(ctx, privilege.CREATEDB, userName)
+				createDB, err := userHasPrivilegeOrRoleOption(ctx, p, userName, options, privilege.CREATEDB)
 				if err != nil {
 					return err
 				}
@@ -3048,7 +3083,7 @@ https://www.postgresql.org/docs/9.5/view-pg-roles.html`,
 					return err
 				}
 
-				replication, err := userHasReplicationPrivilegeOrRoleOption(ctx, p, userName)
+				replication, err := userHasReplicationPrivilegeOrRoleOption(ctx, p, userName, options)
 				if err != nil {
 					return err
 				}
@@ -3062,7 +3097,7 @@ https://www.postgresql.org/docs/9.5/view-pg-roles.html`,
 					tree.MakeDBool(isSuper || tree.DBool(createDB)),   // rolcreatedb
 					tree.DBoolFalse,                                   // rolcatupdate
 					tree.MakeDBool(roleCanLogin),                      // rolcanlogin.
-					tree.MakeDBool(replication),                       // rolreplication
+					tree.MakeDBool(tree.DBool(replication)),           // rolreplication
 					negOneVal,                                         // rolconnlimit
 					passwdStarString,                                  // rolpassword
 					rolValidUntil,                                     // rolvaliduntil

--- a/pkg/sql/rolemembershipcache/BUILD.bazel
+++ b/pkg/sql/rolemembershipcache/BUILD.bazel
@@ -10,6 +10,7 @@ go_library(
         "//pkg/kv/kvpb",
         "//pkg/roachpb",
         "//pkg/security/username",
+        "//pkg/sql/catalog",
         "//pkg/sql/catalog/descpb",
         "//pkg/sql/catalog/descs",
         "//pkg/sql/catalog/lease",

--- a/pkg/sql/rolemembershipcache/cache.go
+++ b/pkg/sql/rolemembershipcache/cache.go
@@ -13,6 +13,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
@@ -33,8 +34,10 @@ import (
 // MembershipCache is a shared cache for role membership information.
 type MembershipCache struct {
 	syncutil.Mutex
-	tableVersion descpb.DescriptorVersion
-	boundAccount mon.BoundAccount
+	roleMembersVersion    descpb.DescriptorVersion
+	roleOptionsVersion    descpb.DescriptorVersion
+	dbRoleSettingsVersion descpb.DescriptorVersion
+	boundAccount          mon.BoundAccount
 	// userCache is a mapping from username to userRoleMembership.
 	userCache map[username.SQLUsername]userRoleMembership
 	// populateCacheGroup ensures that there is at most one request in-flight
@@ -49,6 +52,37 @@ type MembershipCache struct {
 
 // userRoleMembership is a mapping of "rolename" -> "with admin option".
 type userRoleMembership map[username.SQLUsername]bool
+
+// cacheReadTSOptions contains options for RunAtCacheReadTS.
+type cacheReadTSOptions struct {
+	checkRoleOptionsVersion          bool
+	checkDatabaseRoleSettingsVersion bool
+}
+
+// CacheReadTSOption is a functional option for RunAtCacheReadTS.
+type CacheReadTSOption interface {
+	apply(*cacheReadTSOptions)
+}
+
+type cacheReadTSOptionFunc func(*cacheReadTSOptions)
+
+func (f cacheReadTSOptionFunc) apply(o *cacheReadTSOptions) { f(o) }
+
+// CheckRoleOptionsVersion returns an option that causes RunAtCacheReadTS to
+// also check the version of the system.role_options table.
+func CheckRoleOptionsVersion() CacheReadTSOption {
+	return cacheReadTSOptionFunc(func(o *cacheReadTSOptions) {
+		o.checkRoleOptionsVersion = true
+	})
+}
+
+// CheckDatabaseRoleSettingsVersion returns an option that causes RunAtCacheReadTS
+// to also check the version of the system.database_role_settings table.
+func CheckDatabaseRoleSettingsVersion() CacheReadTSOption {
+	return cacheReadTSOptionFunc(func(o *cacheReadTSOptions) {
+		o.checkDatabaseRoleSettingsVersion = true
+	})
+}
 
 // NewMembershipCache initializes a new MembershipCache.
 func NewMembershipCache(
@@ -67,31 +101,93 @@ func NewMembershipCache(
 // table version matches the cached table version, and if it does, db is
 // used to start a separate transaction at the cached timestamp.
 func (m *MembershipCache) RunAtCacheReadTS(
-	ctx context.Context, db descs.DB, txn descs.Txn, f func(context.Context, descs.Txn) error,
+	ctx context.Context,
+	db descs.DB,
+	txn descs.Txn,
+	f func(context.Context, descs.Txn) error,
+	opts ...CacheReadTSOption,
 ) error {
-	tableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleMembersTableID)
+	var options cacheReadTSOptions
+	for _, opt := range opts {
+		opt.apply(&options)
+	}
+
+	roleMembersDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleMembersTableID)
 	if err != nil {
 		return err
+	}
+
+	var roleOptionsDesc, dbRoleSettingsDesc catalog.TableDescriptor
+	if options.checkRoleOptionsVersion {
+		roleOptionsDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleOptionsTableID)
+		if err != nil {
+			return err
+		}
+	}
+	if options.checkDatabaseRoleSettingsVersion {
+		dbRoleSettingsDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.DatabaseRoleSettingsTableID)
+		if err != nil {
+			return err
+		}
 	}
 
 	var readTS hlc.Timestamp
 	func() {
 		m.Lock()
 		defer m.Unlock()
-		if tableDesc.IsUncommittedVersion() {
+
+		// If the option was specified, we need to verify that the cache is aware
+		// of the latest version of the role_options table.
+		defer func() {
+			if roleOptionsDesc != nil {
+				if roleOptionsDesc.IsUncommittedVersion() {
+					readTS = hlc.Timestamp{}
+				} else if roleOptionsDesc.GetVersion() > m.roleOptionsVersion {
+					// Update versions and moved the read timestamp forward. For the
+					// in-progress read, use the current transaction's timestamp.
+					readTS = hlc.Timestamp{}
+					m.roleOptionsVersion = roleOptionsDesc.GetVersion()
+					if m.readTS.Less(roleOptionsDesc.GetModificationTime()) {
+						m.readTS = roleOptionsDesc.GetModificationTime()
+					}
+					return
+				}
+			}
+		}()
+
+		// If the option was specified, we need to verify that the cache is aware
+		// of the latest version of the database_role_settings table.
+		defer func() {
+			if dbRoleSettingsDesc != nil {
+				if dbRoleSettingsDesc.IsUncommittedVersion() {
+					readTS = hlc.Timestamp{}
+				} else if dbRoleSettingsDesc.GetVersion() > m.dbRoleSettingsVersion {
+					// Update versions and moved the read timestamp forward. For the
+					// in-progress read, use the current transaction's timestamp.
+					readTS = hlc.Timestamp{}
+					m.dbRoleSettingsVersion = dbRoleSettingsDesc.GetVersion()
+					if m.readTS.Less(dbRoleSettingsDesc.GetModificationTime()) {
+						m.readTS = dbRoleSettingsDesc.GetModificationTime()
+					}
+					return
+				}
+			}
+		}()
+
+		if roleMembersDesc.IsUncommittedVersion() {
 			return
 		}
-		if tableDesc.GetVersion() > m.tableVersion {
+		if roleMembersDesc.GetVersion() > m.roleMembersVersion {
 			return
 		}
-		if tableDesc.GetVersion() < m.tableVersion {
-			readTS = tableDesc.GetModificationTime()
+		if roleMembersDesc.GetVersion() < m.roleMembersVersion {
+			readTS = roleMembersDesc.GetModificationTime()
 			return
 		}
 		// The cached ts could be from long ago, so use the table modification
 		// if it's more recent.
-		if m.readTS.Less(tableDesc.GetModificationTime()) {
-			readTS = tableDesc.GetModificationTime()
+		if m.readTS.Less(roleMembersDesc.GetModificationTime()) {
+			readTS = roleMembersDesc.GetModificationTime()
 			return
 		}
 		readTS = m.readTS
@@ -116,15 +212,30 @@ func (m *MembershipCache) RunAtCacheReadTS(
 			func() {
 				m.Lock()
 				defer m.Unlock()
-				m.tableVersion = 0
+				m.roleMembersVersion = 0
 			}()
-			txn.Descriptors().ReleaseSpecifiedLeases(ctx, []lease.IDVersion{
+			leasesToRelease := []lease.IDVersion{
 				{
-					Name:    tableDesc.GetName(),
-					ID:      tableDesc.GetID(),
-					Version: tableDesc.GetVersion(),
+					Name:    roleMembersDesc.GetName(),
+					ID:      roleMembersDesc.GetID(),
+					Version: roleMembersDesc.GetVersion(),
 				},
-			})
+			}
+			if roleOptionsDesc != nil {
+				leasesToRelease = append(leasesToRelease, lease.IDVersion{
+					Name:    roleOptionsDesc.GetName(),
+					ID:      roleOptionsDesc.GetID(),
+					Version: roleOptionsDesc.GetVersion(),
+				})
+			}
+			if dbRoleSettingsDesc != nil {
+				leasesToRelease = append(leasesToRelease, lease.IDVersion{
+					Name:    dbRoleSettingsDesc.GetName(),
+					ID:      dbRoleSettingsDesc.GetID(),
+					Version: dbRoleSettingsDesc.GetVersion(),
+				})
+			}
+			txn.Descriptors().ReleaseSpecifiedLeases(ctx, leasesToRelease)
 			return m.RunAtCacheReadTS(ctx, db, txn, f)
 		}
 		return err
@@ -149,7 +260,7 @@ func (m *MembershipCache) GetRolesForMember(
 		return nil, err
 	}
 
-	tableVersion := roleMembersTableDesc.GetVersion()
+	roleMembersVersion := roleMembersTableDesc.GetVersion()
 	if roleMembersTableDesc.IsUncommittedVersion() {
 		return resolveRolesForMember(ctx, txn, member)
 	}
@@ -183,7 +294,7 @@ func (m *MembershipCache) GetRolesForMember(
 				{
 					Name:    roleMembersTableDesc.GetName(),
 					ID:      roleMembersTableDesc.GetID(),
-					Version: tableVersion,
+					Version: roleMembersVersion,
 				},
 				{
 					Name:    systemUsersTableDesc.GetName(),
@@ -205,14 +316,14 @@ func (m *MembershipCache) GetRolesForMember(
 	userMapping, found, refreshCache := func() (userRoleMembership, bool, bool) {
 		m.Lock()
 		defer m.Unlock()
-		if m.tableVersion < tableVersion {
+		if m.roleMembersVersion < roleMembersVersion {
 			// If the cache is based on an old table version, then update version and
 			// drop the map.
-			m.tableVersion = tableVersion
+			m.roleMembersVersion = roleMembersVersion
 			m.userCache = make(map[username.SQLUsername]userRoleMembership)
 			m.boundAccount.Empty(ctx)
 			return nil, false, true
-		} else if m.tableVersion > tableVersion {
+		} else if m.roleMembersVersion > roleMembersVersion {
 			// If the cache is based on a newer table version, then this transaction
 			// should not use the cached data.
 			return nil, false, true
@@ -253,7 +364,7 @@ func (m *MembershipCache) GetRolesForMember(
 	// ensure that we are reading from the right version of the table.
 	newTxnTimestamp := txn.KV().ReadTimestamp()
 	future, _ := m.populateCacheGroup.DoChan(ctx,
-		fmt.Sprintf("refreshMembershipCache-%d", tableVersion),
+		fmt.Sprintf("refreshMembershipCache-%d", roleMembersVersion),
 		singleflight.DoOpts{
 			Stop:               m.stopper,
 			InheritCancelation: false,
@@ -287,7 +398,7 @@ func (m *MembershipCache) GetRolesForMember(
 				// Update membership if the table version hasn't changed.
 				m.Lock()
 				defer m.Unlock()
-				if m.tableVersion != tableVersion {
+				if m.roleMembersVersion != roleMembersVersion {
 					// Table version has changed while we were looking: don't cache the data.
 					return
 				}

